### PR TITLE
fix(perf-issue): query Groups tables when searching for performance issues

### PR DESCRIPTION
**Pull request by barkbarkimashark**
Searching the issue stream for performance issues using a general search term currently doesn't work as expected. 

Screen Shot 2022-11-02 at 11 46 14 AM

Screen Shot 2022-11-02 at 11 49 27 AM

The current behavior queries the `transaction` table and does a `LIKE` match on the `transaction_name` field (or `message` field - they're aliases anyways). Values in `transaction_name` look like the root of where the transaction started, e.g. `'culprit': '/api/0/organizations/{organization_slug}/events/',`, and so searching on `transaction_name` breaks the expected behavior which is to search on the 'culprit which caused the issue' (sentry_groupedmessage.message column). So, for performance issues, instead of querying transactions to find the culprit, we pre-query postgres for the culprit and join the transactions with the Groups.

This PR adds some special handling when a general search term is present. 

Fixes: ISP-49

### CODE REVIEW

File: src/sentry/search/snuba/executors.py [hunk 1]
```diff
@@ -493,6 +494,29 @@ def query(
...
+                                        GroupType.PERFORMANCE_N_PLUS_ONE.value,
+                                        GroupType.PERFORMANCE_SLOW_SPAN.value,
+                                        GroupType.PERFORMANCE_SEQUENTIAL_SLOW_SPANS.value,
+                                        GroupType.PERFORMANCE_LONG_TASK_SPANS.value,
+                                        GroupType.PERFORMANCE_RENDER_BLOCKING_ASSET_SPAN.value,
+                                        GroupType.PERFORMANCE_DUPLICATE_SPANS.value,
+                                        GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES.value,
+                                    )
+                                )
+                                & Q(message__icontains=sf.value.raw_value)
```
Comment (wedamija) [inline src/sentry/search/snuba/executors.py:516]: Nit: You could include this in the previous `Q`, separate terms there are inherently anded together

**Review by wedamija — COMMENTED** [review]
What do we have in the `message` field for transactions? Is it something that we might also want to search on?

**Comment by barkbarkimashark** [comment]
> What do we have in the `message` field for transactions? Is it something that we might also want to search on?

The values look like `sentry.tasks.activity.send_activity_notifications` or `/api/0/organizations/{organization_slug}/events/`, which look like the transaction location or culprit.

**Comment by wedamija** [comment]
> > What do we have in the `message` field for transactions? Is it something that we might also want to search on?
> 
> The values look like `sentry.tasks.activity.send_activity_notifications` or `/api/0/organizations/{organization_slug}/events/`, which look like the transaction location or culprit.

One problem with this is that a search in discover will now work differently to a search in the issue list. Is there any way we can handle both cases? 

I think that when we move forward with doing our own indexing we'll also be able to solve this, since we can store our own title in snuba and then a search on message could search both fields (or we could also insert the title into the message)

**Comment by barkbarkimashark** [comment]
> > > What do we have in the `message` field for transactions? Is it something that we might also want to search on?
> > 
> > 
> > The values look like `sentry.tasks.activity.send_activity_notifications` or `/api/0/organizations/{organization_slug}/events/`, which look like the transaction location or culprit.
> 
> One problem with this is that a search in discover will now work differently to a search in the issue list. Is there any way we can handle both cases?
> 
> I think that when we move forward with doing our own indexing we'll also be able to solve this, since we can store our own title in snuba and then a search on message could search both fields (or we could also insert the title into the message)

Do we need searching on discover to work the same as issue search?

**Comment by wedamija** [comment]
> > > > What do we have in the `message` field for transactions? Is it something that we might also want to search on?
> > > 
> > > 
> > > The values look like `sentry.tasks.activity.send_activity_notifications` or `/api/0/organizations/{organization_slug}/events/`, which look like the transaction location or culprit.
> > 
> > 
> > One problem with this is that a search in discover will now work differently to a search in the issue list. Is there any way we can handle both cases?
> > I think that when we move forward with doing our own indexing we'll also be able to solve this, since we can store our own title in snuba and then a search on message could search both fields (or we could also insert the title into the message)
> 
> Do we need searching on discover to work the same as issue search?

Yeah, ideally a search in discover that turns up a set of transactions would turn up performance issues related to those same transactions.

**Comment by barkbarkimashark** [comment]
> > > > > What do we have in the `message` field for transactions? Is it something that we might also want to search on?
> > > > 
> > > > 
> > > > The values look like `sentry.tasks.activity.send_activity_notifications` or `/api/0/organizations/{organization_slug}/events/`, which look like the transaction location or culprit.
> > > 
> > > 
> > > One problem with this is that a search in discover will now work differently to a search in the issue list. Is there any way we can handle both cases?
> > > I think that when we move forward with doing our own indexing we'll also be able to solve this, since we can store our own title in snuba and then a search on message could search both fields (or we could also insert the title into the message)
> > 
> > 
> > Do we need searching on discover to work the same as issue search?
> 
> Yeah, ideally a search in discover that turns up a set of transactions would turn up performance issues related to those same transactions.

Hmm, I see what you mean. 

What if we concat `Group.metadata.title` + `Group.metadata.location` to form the new Group.message value like what seems to be the case for error Groups? The issue search will potentially return more results compared to discover search since itll have the `N+1 Query` part in the message, but maybe that's fine?

```
'metadata': {
  'title': 'N+1 Query',
  'location': 'sentry.tasks.activity.send_activity_notifications',
  ...
}
```

Screen Shot 2022-11-07 at 4 26 24 PM

**Comment by wedamija** [comment]
> > > > > > What do we have in the `message` field for transactions? Is it something that we might also want to search on?
> > > > > 
> > > > > 
> > > > > The values look like `sentry.tasks.activity.send_activity_notifications` or `/api/0/organizations/{organization_slug}/events/`, which look like the transaction location or culprit.
> > > > 
> > > > 
> > > > One problem with this is that a search in discover will now work differently to a search in the issue list. Is there any way we can handle both cases?
> > > > I think that when we move forward with doing our own indexing we'll also be able to solve this, since we can store our own title in snuba and then a search on message could search both fields (or we could also insert the title into the message)
> > > 
> > > 
> > > Do we need searching on discover to work the same as issue search?
> > 
> > 
> > Yeah, ideally a search in discover that turns up a set of transactions would turn up performance issues related to those same transactions.
> 
> Hmm, I see what you mean.
> 
> What if we concat `Group.metadata.title` + `Group.metadata.location` to form the new Group.message value like what seems to be the case for error Groups? The issue search will potentially return more results compared to discover search since itll have the `N+1 Query` part in the message, but maybe that's fine?
> 
> ```
> 'metadata': {
>   'title': 'N+1 Query',
>   'location': 'sentry.tasks.activity.send_activity_notifications',
>   ...
> }
> ```
> 
> Screen Shot 2022-11-07 at 4 26 24 PM

The problem with that is that then we only have the transaction name from the latest event to search on. So if there are multiple transactions in a single performance issue we'd still have problems. Maybe that's fine as a stopgap for now though?

**Comment by barkbarkimashark** [comment]
> > > > > > > What do we have in the `message` field for transactions? Is it something that we might also want to search on?
> > > > > > 
> > > > > > 
> > > > > > The values look like `sentry.tasks.activity.send_activity_notifications` or `/api/0/organizations/{organization_slug}/events/`, which look like the transaction location or culprit.
> > > > > 
> > > > > 
> > > > > One problem with this is that a search in discover will now work differently to a search in the issue list. Is there any way we can handle both cases?
> > > > > I think that when we move forward with doing our own indexing we'll also be able to solve this, since we can store our own title in snuba and then a search on message could search both fields (or we could also insert the title into the message)
> > > > 
> > > > 
> > > > Do we need searching on discover to work the same as issue search?
> > > 
> > > 
> > > Yeah, ideally a search in discover that turns up a set of transactions would turn up performance issues related to those same transactions.
> > 
> > 
> > Hmm, I see what you mean.
> > What if we concat `Group.metadata.title` + `Group.metadata.location` to form the new Group.message value like what seems to be the case for error Groups? The issue search will potentially return more results compared to discover search since itll have the `N+1 Query` part in the message, but maybe that's fine?
> > ```
> > 'metadata': {
> >   'title': 'N+1 Query',
> >   'location': 'sentry.tasks.activity.send_activity_notifications',
> >   ...
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Screen Shot 2022-11-07 at 4 26 24 PM
> 
> The problem with that is that then we only have the transaction name from the latest event to search on. So if there are multiple transactions in a single performance issue we'd still have problems. Maybe that's fine as a stopgap for now though?

Yeah, this is really a stopgap tbh. I'm not totally sure if users really will query on the performance issue title this is supposed to solve, but its nice to have it consistent with error issues.

File: src/sentry/search/snuba/executors.py [hunk 2]
```diff
@@ -493,6 +494,27 @@ def query(
...
+                            | Q(
+                                type__in=(
+                                    GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES.value,
+                                    GroupType.PERFORMANCE_N_PLUS_ONE.value,
+                                    GroupType.PERFORMANCE_SLOW_SPAN.value,
+                                    GroupType.PERFORMANCE_SEQUENTIAL_SLOW_SPANS.value,
+                                    GroupType.PERFORMANCE_LONG_TASK_SPANS.value,
+                                    GroupType.PERFORMANCE_RENDER_BLOCKING_ASSET_SPAN.value,
+                                    GroupType.PERFORMANCE_DUPLICATE_SPANS.value,
+                                    GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES.value,
```
Comment (rjo100) [inline src/sentry/search/snuba/executors.py:512]: repeat of line `505`

File: src/sentry/search/snuba/backend.py [hunk 3]
```diff
@@ -533,6 +534,26 @@ def _get_queryset_conditions(
...
+                        | Q(
+                            type__in=(
+                                GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES.value,
+                                GroupType.PERFORMANCE_N_PLUS_ONE.value,
+                                GroupType.PERFORMANCE_SLOW_SPAN.value,
+                                GroupType.PERFORMANCE_SEQUENTIAL_SLOW_SPANS.value,
+                                GroupType.PERFORMANCE_LONG_TASK_SPANS.value,
+                                GroupType.PERFORMANCE_RENDER_BLOCKING_ASSET_SPAN.value,
+                                GroupType.PERFORMANCE_DUPLICATE_SPANS.value,
+                                GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES.value,
```
Comment (rjo100) [inline src/sentry/search/snuba/backend.py:550]: duplicate of line `543`

**Review by rjo100 — APPROVED** [review]
approving as long as dan's comments are addressed, which they seem to be

**Pull request merged**